### PR TITLE
dest_{prod,lam}: no Cast case (it's removed by whd)

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

**Review by ppedrot — COMMENTED** [review]
This looks correct to me, but I don't really see the interest. It's not going to give any kind of performance improvement and hardly makes the code more readable.

**Comment by maximedenes** [comment]
Is it clear that whd should remove casts? This can have dramatic effect (think of VMcasts for example). It could be the combinator discriminating the head that ignores casts, instead of reduction removing them. But I'm not completely sure what is the right thing to do.

**Comment by gares** [comment]
A cast `(t : T)`, in spaghetti systems, used to be encoded as `let x : T := t in x`.  Having an extra constructor in terms just for type annotations has benefits (you can attach to it stuff like `VMCast`) but also makes 1 extra case each and every time you write `match t with`.

Given the choice that was made in Coq, I think wiping out casts in *all* reduction strategies is fine.
Ignoring it in all `kind_of_*` (and `dest*`) *should* be done.
Providing a low level API giving access to the cast is then to be provided.
In other words, IMHO, `Cast` is not a constructor of `Constr.t`.  It happens to be there, and hacked in all possible ways.

**Review by gares — APPROVED** [review]
If you read code like this one you wonder "does whd remove casts?".  Given that the answer is YES, you should not waste time wondering.

**Comment by gares** [comment]
> hardly makes the code more readable.

It does. if you see code like `if power x 2 < 0 then ... else stuff` you spend precious brain cycles to wonder why on earth `power` could return a negative number... it can't... (maybe it did, before someone fixed it)
